### PR TITLE
[FLINK-6387] [webfrontend]Flink UI support access log

### PR DESCRIPTION
**Pull request by shijinkui**
Record the use request to the access log. Append use access to the log file.

- [X] General
  - The pull request references the related JIRA issue ("[FLINK-6387] [webfrontend]Flink UI support access log")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [X] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [X] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by StephanEwen** [comment]
I think this is a good issue.
I am wondering if we may want to approach this a bit broader even and define certain "access levels" to the UI:

  - view jobs only
  - view jobs and logs
  - view jobs, logs, cancel jobs
  - view jobs, logs, cancel jobs, submit new jobs

What do you think?

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java [hunk 1]
```diff
@@ -645,6 +645,10 @@
 	@Deprecated
 	public static final String JOB_MANAGER_WEB_LOG_PATH_KEY = "jobmanager.web.log.path";
 
+	/** Config parameter indicating whether jobs can be uploaded and run from the web-frontend. */
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java]: This doc isn't in sync with what the implementation actually does. It doesn't restrict anything but only logs all requests.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/HttpRequestHandler.java [hunk 2]
```diff
@@ -183,4 +194,28 @@ else if (currentDecoder != null && msg instanceof HttpContent) {
...
 		}
 	}
+
+  /**
+   * Record the access log if enable configure of
+   * {@link org.apache.flink.configuration.ConfigConstants#JOB_MANAGER_WEB_ACCESSLOG_ENABLE}.
+   * record format:
+   * remote_addr - [time_local] "request_method URI protocolVersion" "http_referer" "http_user_agent"
+   */
+	private void accesslog(ChannelHandlerContext ctx, HttpRequest req) {
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/HttpRequestHandler.java]: this should be renamed to `logAccess`.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/HttpRequestHandler.java [hunk 3]
```diff
@@ -183,4 +194,28 @@ else if (currentDecoder != null && msg instanceof HttpContent) {
...
+		HttpHeaders headers = req.headers();
+		if (headers != null) {
+			String line = ctx.channel().remoteAddress() + " - [" + new Date() + "] \""
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/HttpRequestHandler.java]: Please merge this string generation into the log statement.

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java [hunk 4]
```diff
@@ -645,6 +645,10 @@
...
+	public static final ConfigOption<Boolean> JOB_MANAGER_WEB_ACCESSLOG_ENABLE =
+			key("jobmanager.web.accesslog.enable").defaultValue(false);
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java]: please move the `.defaultValue` call into a new line. It should also be moved to the JobManagerOptions class.

**Comment by StephanEwen** [comment]
@zentol @shijinkui Ah, forget what I said, I only read the description and thought it was something different ;-)

I thought this was about forbidding UI users to access the logs (which seems valid since the logs may contain sensitive information).

This seems to be about auditing, so completely different...

**Comment by shijinkui** [comment]
> I can see this flooding the logs like crazy, especially with things like metrics and watermarks that update often. Some kind of filtering is probably necessary here.

You are right. The crazy log can separated into a new log file.

> This seems to be about auditing, so completely different...

The main purpose is record every access, like auditing log.

**Comment by shijinkui** [comment]
@StephanEwen @zentol   The Flink webmonitor is netty inside. In generally, Netty is used as a tcp transfer. For a web or API server, jetty or Playframework maybe more suitable. Especially we must fill much gaps and holes on web security if choose netty as web framework.

I think we should re-consider whether Netty is suitable?  Current Flink UI is hard to extend.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/HttpRequestHandler.java [hunk 5]
```diff
@@ -197,17 +197,16 @@ else if (currentDecoder != null && msg instanceof HttpContent) {
...
+   * {@link org.apache.flink.configuration.JobManagerOptions#JOB_MANAGER_WEB_ACCESSLOG_ENABLE}.
    * record format:
    * remote_addr - [time_local] "request_method URI protocolVersion" "http_referer" "http_user_agent"
    */
-	private void accesslog(ChannelHandlerContext ctx, HttpRequest req) {
+	private void logAccess(ChannelHandlerContext ctx, HttpRequest req) {
 		HttpHeaders headers = req.headers();
 		if (headers != null) {
-			String line = ctx.channel().remoteAddress() + " - [" + new Date() + "] \""
+			LOG.info(ctx.channel().remoteAddress() + " - [" + new Date() + "] \""
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/HttpRequestHandler.java]: Please rework this to use placeholders instead.

**Comment by zentol** [comment]
Subsumed in FLINK-7040.

**Pull request closed without merging**